internal/tools/aws/database_multitenant.go

```diff
@@ -0,0 +1,730 @@
+
+	"github.com/mattermost/mattermost-cloud/model"
+	mmv1alpha1 "github.com/mattermost/mattermost-operator/pkg/apis/mattermost/v1alpha1"
+
+	// MySQL implementation of database/sql
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// SQLDatabaseManager is an interface that describes operations to query and to
+// close connection with a database. It's used mainly to implement a client that
+// needs to perform non-complex queries in a SQL database instance.
+type SQLDatabaseManager interface {
+	QueryContext(ctx context.Context, query string, args ...interface{}) (*sql.Rows, error)
+	Close() error
+}
+
+// RDSMultitenantDatabase is a database backed by RDS that supports multi-tenancy.
+type RDSMultitenantDatabase struct {
+	client         *Client
+	db             SQLDatabaseManager
+	installationID string
+}
+
+// NewRDSMultitenantDatabase returns a new instance of RDSMultitenantDatabase that implements database interface.
+func NewRDSMultitenantDatabase(installationID string, client *Client) *RDSMultitenantDatabase {
+	return &RDSMultitenantDatabase{
+		client:         client,
+		installationID: installationID,
+	}
+}
+
+// Teardown removes all AWS resources related to a RDS multitenant database.
+func (d *RDSMultitenantDatabase) Teardown(store model.InstallationDatabaseStoreInterface, keepData bool, logger log.FieldLogger) error {
+	databaseName := MattermostRDSDatabaseName(d.installationID)
+
+	logger = logger.WithField("rds-multitenant-database", databaseName)
+	logger.Info("Tearing down RDS database and database secret")
+
+	multitenantDatabases, err := store.GetMultitenantDatabases(&model.MultitenantDatabaseFilter{
+		InstallationID: d.installationID,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "cannot teardown multitenant database for installation ID %s", d.installationID)
+	}
+	if len(multitenantDatabases) > 1 {
+		return errors.Errorf("more than one multitenant database per installation is not supported by this database (found %d)", len(multitenantDatabases))
+	}
+	if len(multitenantDatabases) == 0 {
+		logger.Warnf("Installation ID %s does not belong to any of the multitenant databases. Nothing to be torn down.", d.installationID)
+
+		return nil
+	}
+
+	err = d.removeRDSDatabase(multitenantDatabases[0], databaseName, store, logger)
+	if err != nil {
+		return errors.Wrapf(err, "unable to finish teardown of multitenant database name %s", databaseName)
+	}
+
+	multitenantDatabaseSecretName := RDSMultitenantSecretName(d.installationID)
+
+	_, err = d.client.Service().secretsManager.DeleteSecret(&secretsmanager.DeleteSecretInput{
+		SecretId: aws.String(multitenantDatabaseSecretName),
+	})
+	if err != nil && !IsErrorCode(err, secretsmanager.ErrCodeResourceNotFoundException) {
+		return errors.Wrapf(err, "failed to delete multitenant database secret name %s", multitenantDatabaseSecretName)
+	}
```

Comment:
All of this happens after the `MultitenantStore` has been updated to remove the installation ID. Due to this, there is the possibility to leak the following secret. The store should be the final thing to be updated.

Reply:
Unless the installation ID is gone, I don't see how. Secrets are managed by AWS and we just reference them by a name that depends on the installations ID. It's completely independent of the provisioner's persistent data.

Reply:
Right, so if the installation ID is removed on the first pass on line 80, but then we hit an error here on line 82 it will fail and eventually retry.

When it does, this check above will return from this function and the secret cleanup will be missed.

```
	if len(multitenantDatabases) == 0 {
		logger.Warnf("Installation ID %s does not belong to any of the multitenant databases. Nothing to be torn down.", d.installationID)

		return nil
	}
```

Reply:
Oh right, I forgot that I added this check. Problem of moving it before `removeRDSDatabase()` is that I need the secrets in oder to drop the database. I fix the logic.

Reply:
I'm still writing the tests for the provisioning part. I will make sure that this is being tested.